examples/beacon-api-sidecar-fetcher/src/main.rs

```diff
@@ -0,0 +1,235 @@
+//! Run with
+//!
+//! ```not_rust
+//! cargo run -p beacon-api-sidecar-fetcher -- node
+
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use thiserror::Error;
+
+use futures_util::{stream::FuturesUnordered, Future, Stream, StreamExt};
+use reqwest::{Error, StatusCode};
+use reth::{
+    primitives::{alloy_primitives::TxHash, BlobTransactionSidecar, BlockHash, BlockNumber},
+    providers::CanonStateNotification,
+    rpc::types::engine::BlobsBundleV1,
+    transaction_pool::{BlobStoreError, TransactionPoolExt},
+};
+
+#[tokio::main]
+async fn main() -> eyre::Result<()> {
+    Ok(())
+}
+
+//TODO Add nicer errors.
+#[derive(Debug, Error)]
+pub enum SideCarError {
+    #[error("Reqwest encountered an error: {0}")]
+    ReqwestError(Error),
+
+    #[error("There was an error grabbing the blob from the tx pool: {0}")]
+    TransactionPoolError(BlobStoreError),
+
+    #[error("400: {0}")]
+    InvalidBlockID(String),
+
+    #[error("404: {0}")]
+    BlockNotFound(String),
+
+    #[error("500: {0}")]
+    InternalError(String),
+
+    #[error("Network error: {0}")]
+    NetworkError(String),
+
+    #[error("Data parsing error: {0}")]
+    DeserializationError(String),
+
+    #[error("{0} Error: {1}")]
+    UnknownError(u16, String),
+}
+
+#[derive(Debug, Clone)]
+pub struct MinedSidecar {
+    tx_hash: TxHash,
+    block_hash: BlockHash,
+    block_number: BlockNumber,
+    sidecar: BlobTransactionSidecar,
+}
+
+#[derive(Debug)]
+pub struct MinedSidecarStream<St, P>
+where
+    St: Stream<Item = CanonStateNotification> + Send + Unpin + 'static,
+{
+    events: St,
+    pool: P,
+    client: reqwest::Client,
+    pending_requests:
+        FuturesUnordered<Pin<Box<dyn Future<Output = Result<Vec<MinedSidecar>, SideCarError>>>>>, /* TODO make vec */
+    queued_actions: VecDeque<MinedSidecar>, /* Buffer for
+                                             * ready items */
+}
+
+impl<St, P> Stream for MinedSidecarStream<St, P>
+where
+    St: Stream<Item = CanonStateNotification> + Send + Unpin + 'static,
+    P: TransactionPoolExt + Unpin + 'static,
+{
+    type Item = Result<MinedSidecar, SideCarError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this: &mut MinedSidecarStream<St, P> = self.get_mut();
+
+        // return any buffered result
+        if let Some(mined_sidecar) = this.queued_actions.pop_front() {
+            return Poll::Ready(Some(Ok(mined_sidecar)));
+        }
```

Comment:
this needs an outer loop, so we continue here if we push to the buffer

Reply:
Done